cpp/include/cudf/detail/gather.cuh

```diff
@@ -0,0 +1,256 @@
+#include <cudf/cudf.h>
+#include <cudf/types.hpp>
+#include <cudf/copying.hpp>
+#include <utilities/bit_util.cuh>
+#include <utilities/cudf_utils.h>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/null_mask.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_device_view.cuh>
+
+#include <rmm/thrust_rmm_allocator.h>
+
+#include <algorithm>
+
+#include <thrust/functional.h>
+#include <thrust/logical.h>
+#include <thrust/gather.h>
+#include <thrust/iterator/transform_iterator.h>
+
+
+#include <cub/cub.cuh>
+#include <utilities/column_utils.hpp>
+#include <utilities/cuda_utils.hpp>
+#include <utilities/release_assert.cuh>
+
+#include <bitmask/legacy/valid_if.cuh>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+/**---------------------------------------------------------------------------*
+ * @brief Function object to check if an index is within the bounds [begin,
+ * end).
+ *
+ *---------------------------------------------------------------------------**/
+template <typename map_type>
+struct bounds_checker {
+  size_type begin;
+  size_type end;
+
+  __device__ bounds_checker(size_type begin_, size_type end_)
+    : begin{begin_}, end{end_} {}
+
+  __device__ bool operator()(map_type const index) {
+    return ((index >= begin) && (index < end));
+  }
+};
+
+
+template <bool ignore_out_of_bounds, typename map_iterator_type>
+__global__ void gather_bitmask_kernel(table_device_view source_table,
```

Comment:
@jrhemstad Would appreciate if you could pay special attention to this kernel when reviewing, as I got rid of a substantial amount of code here

Reply:
I'm always a big fan of getting rid of lots of code!